src/module_3/push_exploration.md

```diff
@@ -0,0 +1,5364 @@
+    
+        accuracy                           0.99    216395
+       macro avg       0.49      0.50      0.50    216395
+    weighted avg       0.97      0.99      0.98    216395
+    
+    Confusion Matrix:
+    Watch out: Rows are actual values (N and P), cols are predicted (N and P)
+    [[213182      0]
+     [  3213      0]]
+
+
+    /home/dan1dr/.cache/pypoetry/virtualenvs/zrive-ds-UEx3J_CK-py3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
+      _warn_prf(average, modifier, msg_start, len(result))
+    /home/dan1dr/.cache/pypoetry/virtualenvs/zrive-ds-UEx3J_CK-py3.11/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
+      _warn_prf(average, modifier, msg_start, len(result))
+
+
+Model does not converge with the default number of iterations = 100. Model didn't classified anything as positive! If we try to plot the curves, we will appreciate the classifier it is not a better option than flipping a coin and estimating.
+
+
+```python
+from sklearn.metrics import roc_curve, auc, precision_recall_curve
+
+# Calculate probabilities for the positive class
+y_val_probs = model.predict_proba(X_val)[:, 1]
+
+# ROC Curve
+fpr_val, tpr_val, _ = roc_curve(y_val, y_val_probs)
+roc_auc_val = auc(fpr_val, tpr_val)  # Calculate AUC
+
+plt.figure(figsize=(10,5))
+
+plt.subplot(1, 2, 1)
+plt.plot(fpr_val, tpr_val, label=f'ROC (AUC = {roc_auc_val:.2f})')
+plt.plot([0, 1], [0, 1], color='navy', linestyle='--')
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('ROC')
+plt.legend(loc="lower right")
+
+```
+
+
+
+
+    <matplotlib.legend.Legend at 0x7fca6b239d10>
+
+
+
+
+    
+![png](push_exploration_files/push_exploration_32_1.png)
+    
+
+
+
+```python
+# Precision-Recall Curve
+precision_val, recall_val, _ = precision_recall_curve(y_val, y_val_probs)
+
+plt.subplot(1, 2, 2)
+plt.plot(recall_val, precision_val)
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('Recall')
+plt.ylabel('Precision')
+plt.title('Precision-Recall Curve')
+
+plt.tight_layout()
+plt.show()
+```
+
+
+    
+![png](push_exploration_files/push_exploration_33_0.png)
+    
+
+
+We will use the feature scaling provided by scikit in order to standardized the features and enable the estimator to learn from them easier. This will substract the mean from the feature and divide by the std.
+
+
+```python
+from sklearn.preprocessing import StandardScaler
+
+X = final_data.drop(columns='outcome')
+y = final_data['outcome']
+
+# Feature scaler and apply to X data
+scaler = StandardScaler()
+X = scaler.fit_transform(X)
```

Comment:
this is an important mistake: all the transformations which uses things about the data itself to transform it (here `StandardScaler ` uses the mean and the std of the data) need to be `fit` **only** in the train test. Meaning that in this case, you need to compute the `mean` and `std` **only** from the training, and then you can apply the transformation to the `test` / `eval`. Imagine for example that `train` and `test` don't have the same distribution: that would mean that their means may be different, however you would be computing a common mean by doing the `fit_transform` in `X` instead of `X_train`